### PR TITLE
feat: add retry and delivery confirmation for plivo

### PR DESCRIPTION
**Pull request by sujoy-pal144**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](../../CONTRIBUTING.md#submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [x] I have performed preliminary testing using the [test suite](../../apps/api/OsmoX.postman_collection.json) to ensure that any existing features are not impacted and any new features are working as expected.
- [ ] I have updated the required [api docs](../../apps/api/docs/) as applicable.
- [ ] I have added/updated test cases to the [test suite](../../apps/api/OsmoX.postman_collection.json) as applicable

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login endpoint`)
- [x] Description has been added
- [ ] Related changes have been added (optional)
- [ ] Screenshots have been added (optional)
- [ ] Query request and response examples have been added (as applicable, in case added or updated)
- [ ] Documentation changes have been listed (as applicable)
- [ ] Test suite output is added (as applicable)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

---

**Description:**

Retry and delivery confirmation for plivo

**Related changes:**

N/A

**Screenshots:**

N/A

**Query request and response:**

N/A

**Documentation changes:**

N/A

**Test suite output:**

N/A

**Pending actions:**

N/A

**Additional notes:**

N/A

## Summary by CodeRabbit

- **New Features**
  - Introduced a new interface for enhanced message status response data.
  - Added a new method for retrieving delivery status from the Plivo API.
  - Added a `retry_count` column to track notification retries.
  
- **Enhancements**
  - Updated delivery status constants to include `AWAITING_CONFIRMATION`.
  - Enhanced error handling and status updates in various notification processing methods.

- **Bug Fixes**
  - Refined the naming convention for message UUID fields in the Plivo service.

- **Infrastructure**
  - Implemented database migrations for new columns and updated delivery status values.

### CODE REVIEW

File: apps/api/src/database/migrations/1719299457601-AddRetryCountColumn.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+export class AddRetryCountColumn1719299457601 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.addColumn('notify_notifications', {
+      name: 'retry_count',
+      type: 'int',
+      length: '3',
+      default: 0,
+    } as TableColumn);
+    // Need to handle migration for setting proper delivery_status
+  }
```
[the remark this answers is not in the document]
Comment (sujoy-pal144) [reply apps/api/src/database/migrations/1719299457601-AddRetryCountColumn.ts]: These is outdated suggestion, I have modified the code.

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 2]
```diff
@@ -39,4 +57,14 @@
...
     return response;
   }
+  async getDeliveryStatus(
+    messageUuid: string,
+    providerId: number,
+  ): Promise<PlivoMessageStatusResponse> {
+    await this.assignTransport(providerId);
+
+    const response = await this.plivoClient.messages.get(messageUuid);
+    console.log(response);
```
[the remark this answers is not in the document]
Comment (sujoy-pal144) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts]: These is outdated suggestion, I have modified the code.

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 3]
```diff
@@ -13,6 +13,24 @@
...
+  message_type: string;
+  message_uuid: string;
+  resource_uri: string;
+  to_number: string;
+  total_amount: string;
+  total_rate: string;
+  units: number;
+  status: string;
+  // add any other fields you expect in the response
+}
```
[the remark this answers is not in the document]
Comment (sujoy-pal144) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts]: These is outdated suggestion, I have modified the code.

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 4]
```diff
@@ -39,4 +57,14 @@ export class SmsPlivoService {
...
+  async getDeliveryStatus(
+    messageUuid: string,
+    providerId: number,
+  ): Promise<PlivoMessageStatusResponse> {
+    await this.assignTransport(providerId);
+
+    const response = await this.plivoClient.messages.get(messageUuid);
+    //console.log(response);
+    return response;
+  }
```
[the remark this answers is not in the document]
Comment (sujoy-pal144) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts]: These is outdated suggestion, I have modified the code.

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 5]
```diff
@@ -10,7 +10,25 @@ export interface SmsPlivoData {
...
+  message_time: string;
+  message_type: string;
+  message_uuid: string;
+  resource_uri: string;
+  to_number: string;
+  total_amount: string;
+  total_rate: string;
+  units: number;
+  status: string;
+  // add any other fields you expect in the response
```
[the remark this answers is not in the document]
Comment (sujoy-pal144) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts]: These is outdated suggestion, I have modified the code.

File: apps/api/src/jobs/consumers/notifications/smsPlivo-notifications.job.consumer.ts [hunk 6]
```diff
@@ -21,13 +26,34 @@ export class SmsPlivoNotificationsConsumer extends NotificationConsumer {
...
+          deliveryStatus === 'failed' ||
+          deliveryStatus === 'undelivered' ||
+          deliveryStatus === 'rejected'
+        ) {
+          return { result, deliveryStatus: DeliveryStatus.FAILED };
+        }
+
+        return { result, deliveryStatus: DeliveryStatus.SUCCESS };
+      });
+    }
```
[the remark this answers is not in the document]
Comment (sujoy-pal144) [reply apps/api/src/jobs/consumers/notifications/smsPlivo-notifications.job.consumer.ts]: These is outdated suggestion, I have modified the code.

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 7]
```diff
@@ -10,7 +10,25 @@ export interface SmsPlivoData {
 export interface SmsPlivoResponseData {
   api_id: string;
   message: string;
-  message_uuid: string[];
```
Comment (LakshayaT) [inline apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts:13]: Is the response value for message uuid not an array of string?

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 8]
```diff
@@ -10,7 +10,25 @@ export interface SmsPlivoData {
...
+  messageUuid: string;
+}
+
+export interface PlivoMessageStatusResponse {
+  api_id: string;
+  error_code: string;
+  from_number: string;
+  message_direction: string;
+  messageState: string;
```
Comment (LakshayaT) [inline apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts:21]: Why is this in camelCase when others are in snake_case?

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 9]
```diff
@@ -10,7 +10,25 @@ export interface SmsPlivoData {
...
+  message_time: string;
+  message_type: string;
+  message_uuid: string;
+  resource_uri: string;
+  to_number: string;
+  total_amount: string;
+  total_rate: string;
+  units: number;
+  status: string;
+  // add any other fields you expect in the response
```
Comment (LakshayaT) [inline apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts:31]: Do we expect any other fields here?

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts [hunk 10]
```diff
@@ -27,16 +45,34 @@ export class SmsPlivoService {
...
+  async getDeliveryStatus(
+    messageUuid: string,
+    providerId: number,
+  ): Promise<PlivoMessageStatusResponse> {
+    try {
+      await this.assignTransport(providerId);
 
-    return response;
+      const response = await this.plivoClient.messages.get(messageUuid);
+      //throw new Error(`Custom error for response: ${JSON.stringify(response)}`);
```
Comment (LakshayaT) [inline apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts:72]: Is this needed? Can remove it if not

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts (hunk 8 above)
Comment (LakshayaT) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts:21]: This should be camelCase for all properties (docs don't mention it for messaging, but this reference tells that Node.js uses camelCase: https://docs.plivo.com/docs/voice/api/response/

Addressed in https://github.com/OsmosysSoftware/osmo-x/pull/261/commits/5ce4e67bd9877772054c24dac5a90f5ae9c00aad

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts (hunk 9 above)
Comment (LakshayaT) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts]: Addressed in https://github.com/OsmosysSoftware/osmo-x/pull/261/commits/5ce4e67bd9877772054c24dac5a90f5ae9c00aad based on test run and response obtained

File: apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts (hunk 10 above)
Comment (LakshayaT) [reply apps/api/src/modules/providers/sms-plivo/sms-plivo.service.ts]: Addressed in https://github.com/OsmosysSoftware/osmo-x/pull/261/commits/5ce4e67bd9877772054c24dac5a90f5ae9c00aad

**Review by xixas — APPROVED** [review]

**Pull request merged**